services/slam/builtin/cartographer_int_test.go

```diff
@@ -156,8 +176,136 @@ func TestCartographerIntegration(t *testing.T) {
 
 	// Close out slam service
 	test.That(t, utils.TryClose(context.Background(), svc), test.ShouldBeNil)
+
+	// Don't clear out the directory, since we will re-use the maps for the next run
+	closeOutSLAMService(t, "")
+
+	// added sleep to ensure cartographer stops
+	time.Sleep(time.Millisecond * cartoSleepMS)
+
+	// Remove existing pointclouds, but leave maps and config (so we keep the lua files).
+	test.That(t, resetFolder(name+"/data"), test.ShouldBeNil)
+	// Count the initial number of maps in the map directory
+	numMaps, err := ioutil.ReadDir(name + "/map/")
+	test.That(t, err, test.ShouldBeNil)
+	// Test online mode using the map generated in the offline test
+	t.Log("Testing online mode in localization mode")
+
+	mapRate = 0
+
+	attrCfg = &builtin.AttrConfig{
+		Sensors: []string{"cartographer_int_lidar"},
+		ConfigParams: map[string]string{
+			"mode": "2d",
+			"v":    "1",
+		},
+		MapRateSec:    &mapRate,
+		DataDirectory: name,
+	}
+
+	// Release point cloud for service validation
+	cartographerIntLidarReleasePointCloudChan <- 1
+	// Create slam service using a real cartogapher binary
+	svc, err = createSLAMService(t, attrCfg, "cartographer", golog.NewTestLogger(t), true, true)
+	test.That(t, err, test.ShouldBeNil)
+
+	// Make sure we initialize from a saved map
+	logReader = svc.(internal.Service).GetSLAMProcessBufferedLogReader()
+	for {
+		line, err := logReader.ReadString('\n')
+		test.That(t, err, test.ShouldBeNil)
+		if strings.Contains(line, "Running in localization only mode") {
+			break
+		}
+		test.That(t, strings.Contains(line, "Running in updating mode"), test.ShouldBeFalse)
+		test.That(t, strings.Contains(line, "Running in mapping mode"), test.ShouldBeFalse)
+	}
+
+	// Release point cloud, since cartographer looks for the second most recent point cloud
+	cartographerIntLidarReleasePointCloudChan <- 1
+	for i := 0; i < numCartographerPointClouds-2; i++ {
+		t.Logf("Find log line for point cloud %v", i)
+		cartographerIntLidarReleasePointCloudChan <- 1
+		for {
+			line, err := logReader.ReadString('\n')
+			test.That(t, err, test.ShouldBeNil)
+			if strings.Contains(line, "Passed sensor data to SLAM") {
+				break
+			}
+		}
+	}
+
+	testCartographerPositionAndMap(t, svc)
+
+	// Close out slam service
+	test.That(t, utils.TryClose(context.Background(), svc), test.ShouldBeNil)
+
+	// Test that no new maps were generated
+	numMapsLocalize, err := ioutil.ReadDir(name + "/map/")
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, len(numMapsLocalize), test.ShouldEqual, len(numMaps))
```

Comment:
[minor] considering adding in a data_file check to each one that either checks that no new data was added, new data WAS added and/or how many files should be expected

Reply:
i do not think this is necessary, since we iterate through a fixed number of pcd scans already in the above `for` loop